nemo/utils/model_utils.py

```diff
@@ -0,0 +1,232 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import os
+from pathlib import Path
+
+from omegaconf import DictConfig, ListConfig, OmegaConf
+
+from nemo import logging
+
+
+def resolve_filepath_from_cfg(cfg: DictConfig) -> str:
+    """
+    Parses items of the provided sub-config to find the first potential key that
+    resolves to an existing file or directory.
+
+    NOTE:
+    It <can> potentially mismatch if there exist more than 2 valid paths, and the
+    first path does *not* resolve the the path of the data file (but does resolve to
+    some other valid path).
+
+    To avoid this side-effect, place the data path as the first item on the config file.
+
+    Args:
+        cfg: DictConfig (Sub-config) that should be parsed.
+
+    Returns:
+        A str representing the `key` of the config which hosts the filepath(s),
+        or None in case path could not be resolved.
+    """
+    if hasattr(cfg, 'manifest_filepath'):
+        return 'manifest_filepath'
+
+    for key, value in cfg.items():
+        if type(value) in [list, tuple, ListConfig]:
+            # Count the number of valid paths in the list
+            values_are_paths = 0
+            for val_i in value:
+                val_i = str(val_i)
+
+                if os.path.exists(val_i) or os.path.isdir(val_i):
+                    values_are_paths += 1
+                else:
+                    # reset counter and break inner loop
+                    break
+
+            if values_are_paths == len(value):
+                return key
```

Comment:
[QUESTION]
Do I get it right that you are assuming that:
"if element on a list is a string AND it is a valid path/dir" -> "it is a manifest_filepath

Reply:
Yes, I assume there exists at least 1 key:value pair in the config subset that resolves to one or more correct filepaths/dirs. This loop will select the 1st such key:value pair, and return the key.

This assumption is broken by a few NLP configs (they have one data path stored inside "model" level, not "model.validation_ds" level. This is legal and allowed, the code will backoff resolution and assume your code will handle everything properly. As such, the NLP tests pass without any modifications from my side.

The file name does not need to be `manifest_filepath` - thats why the loop makes no such assumption, and simply tests all values for correct filepaths. (I do have a fastpath for manifest_filepath since we can make immediate assumption without this loop).

Reply:
Can we add control over the key name, set manifest_filepath as default but potentially provide a different name, or call it smth more general, like file_paths. What would this check return if my config contains checkpoint_path: path_to_ckpt as the first key this parsing encounters?

Reply:
@ekmb Good idea. Right now, it will parse all of the keys one by one in order defined in yaml, and the first key it finds that has a valid path, it will assume it is the data path. 

Since it assumes first path is the data path, if somehow checkpoint path is above data_path, manifest_path or file_path (however one devices to call it), then the check will assume that the checkpoint path is the data path and fail later on (since your code will not be able to parse a checkpoint path as if its a data/manifest path).

Ofcourse, if the data path is at the top, doesnt matter what its called, it will be resolved correctly.

Reply:
@ekmb I replaced `manifest_filepath` with `ds_name`. ds_name can be added either via hydra + syntax or by using yaml. Its behaviour is just like manifest_filepath (without the connotation of files or paths) and it will be resolved with priority. 

So if you have checkpoint_path as well as ds_name, then ds_name will be picked even if its lower on the list inside yaml.